iTwin Reality Data integration

jjspace opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

This PR is building off of #12289 to add an easy function to load Reality Data meshes. Currently we are only supporting 3D Tiles meshes and pointcloud types of Reality Data to keep the initial integration simple. More can be added in the future but they will require different datasources and constructors in CesiumJS and will require a little more thought on how to design the API.

<!-- Describe your changes in detail -->

<!-- Consider: Why is this change required? What problem does it solve? -->

<!-- Include screenshots if appropriate -->

## Issue number and link

No issue

<!-- If it fixes an open issue, link to the issue here -->

<!-- Consider: If suggesting a new feature or change, discuss it in an issue first. -->

## Testing plan

- Running the test sandcastle
    - Load the [new sandcastle](http://localhost:8080/Apps/Sandcastle/index.html?src=iTwin%20Demo.html) and make sure it loads.
    - Check that the reality mesh in that iTwin loads and shows in the surrounding area.
<!-- Describe in detail how you tested your changes. If this fixes a bug, list the steps to reproduce the original issue. -->

# Author checklist

- [ ] I have submitted a Contributor License Agreement
- [ ] I have added my name to `CONTRIBUTORS.md`
- [ ] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [ ] I have updated the inline documentation, and included code examples where relevant
- [ ] I have performed a self-review of my code

ggetz reviewed (changes requested):
Overall API looks good @jjspace! Just a few comments.

jjspace commented on packages/engine/Source/Core/ITwinPlatform.js:
```diff
@@ -37,6 +37,64 @@ ITwinPlatform.ExportType = Object.freeze({
+  Other: "Other",
+});
+
+ITwinPlatform.SupportedRealityDataTypes = [
```
I shifted this supported list into the only function that uses it. `RealityDataType` cannot be made private if I want to set it as the type of the param in `createTilesetForRealityDataId` because of the way types get generated and skipped for `private`

jjspace commented:
@ggetz I've added tests now and I think addressed all your comments above. Should be ready for another look and I think it's close to done if not already there

ggetz commented on packages/engine/Source/Scene/ITwinData.js:
```diff
@@ -71,4 +72,63 @@ ITwinData.createTilesetFromIModelId = async function (iModelId, options) {
+ *
+ * If the <code>type</code> or <code>rootDocument</code> are not provided this function
+ * will first request the full metadata for the specified reality data to fill these values.
+ *
```
For consistency with `createTilesetFromIModelId`.

```suggestion
 *
 * @experimental This feature is not final and is subject to change without Cesium's standard deprecation policy.
```

ggetz reviewed (changes requested):
This is looking good @jjspace!

I noticed that the reality mesh looks a bit blurry at the default settings. I was getting a comparable visual to other photogrammetry tilesets after setting the max SSE to `4` —

```
realityMesh.maximumScreenSpaceError = 4;
```

I'm guessing this may be because the geometric error of the tileset is balanced against the iTwin viewer, what may have different tile selection metrics. Have you seen the same for other Reality Data meshes?

What do you think of passing a default of `4` for all with the type of `RealityMesh3DTiles`?

jjspace commented:
@ggetz I agree the tileset in the sandcastle was too low quality by default. I check some other reality meshes I have access to and sse of 4 looked better on all of them too so I just set it as the default as you suggested.

ggetz reviewed (approved):
This looks fantastic, thanks @jjspace!

matmarchand commented on packages/engine/Source/Scene/ITwinData.js:
```diff
@@ -71,4 +72,67 @@ ITwinData.createTilesetFromIModelId = async function (iModelId, options) {
+  iTwinId,
+  realityDataId,
+  type,
+  rootDocument,
```
I believe type and rootDocument should be extracted from the reality data metadata to ensure that they are consistent with what actually exist in the storage container... or maybe this is our way to avoid a call to get metadata since they were already extracted and assumed consistent?

jjspace replied:
> or maybe this is our way to avoid a call to get metadata since they were already extracted and assumed consistent?

@matmarchand You are correct. If either of these values is missing we will fetch the metadata needed to load the tileset. This will likely only the rootDocument commonly but we grab and update both from the new request anyway.

However if the application using this function already has these values we can skip that network call. This could happen if the app [loads the list of reality data](https://developer.bentley.com/apis/reality-management/operations/get-all-reality-data/) using `return=representation` instead of `return=minimal`. Or if the app already called the individual metadata route themselves. No need to waste extra time and bandwidth if the info is already accessible
